core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java

```diff
@@ -270,9 +277,82 @@ default int get(PageId pageId, int pageOffset, int bytesToRead, byte[] buffer, i
    * @param cacheContext cache related context
    * @return number of bytes read, 0 if page is not found, -1 on errors
    */
-  int get(PageId pageId, int pageOffset, int bytesToRead, PageReadTargetBuffer buffer,
+  int get(PageId pageId, int pageOffset, int bytesToRead, ReadTargetBuffer buffer,
       CacheContext cacheContext);
 
+  /**
+   * Reads a part of a page with metrics recording.
+   *
+   * @param pageId page identifier
+   * @param pageOffset offset into the page
+   * @param bytesToRead number of bytes to read in this page
+   * @param buffer destination buffer to write
+   * @param cacheContext cache related context
+   * @param stopwatch stop watch to record metrics
+   * @return number of bytes read, 0 if page is not found, -1 on errors
+   */
+  default int getWithMetrics(PageId pageId, int pageOffset, int bytesToRead,
+      ReadTargetBuffer buffer, CacheContext cacheContext, Stopwatch stopwatch) {
+    stopwatch.reset().start();
+    int bytesRead =
+        get(pageId, pageOffset, bytesToRead, buffer, cacheContext);
+    stopwatch.stop();
+    if (bytesRead > 0) {
+      MetricsSystem.meter(MetricKey.CLIENT_CACHE_BYTES_READ_CACHE.getName()).mark(bytesRead);
+      if (cacheContext != null) {
+        cacheContext.incrementCounter(MetricKey.CLIENT_CACHE_BYTES_READ_CACHE.getMetricName(), BYTE,
+            bytesRead);
+        cacheContext.incrementCounter(
+            MetricKey.CLIENT_CACHE_PAGE_READ_CACHE_TIME_NS.getMetricName(), NANO,
+            stopwatch.elapsed(TimeUnit.NANOSECONDS));
+      }
+    }
+    return bytesRead;
+  }
+
+  /**
+   * Reads a part of a page if the queried page is found in the cache, stores the result in buffer.
+   * Loads the page otherwise.
+   *
+   * @param pageId page identifier
+   * @param pageOffset offset into the page
+   * @param bytesToRead number of bytes to read in this page
+   * @param buffer destination buffer to write
+   * @param cacheContext cache related context
+   * @param stopwatch stop watch to record metrics
+   * @param externalDataSupplier the external data supplier to read a page
+   * @return number of bytes read, 0 if page is not found, -1 on errors
+   */
+  default int getAndLoad(PageId pageId, int pageOffset, int bytesToRead,
```

Comment:
Does this have to have a default implementation? I see in `LocalCacheMagaer` you need to acquire a write lock to synchronize the reading from cache and then loading from fallback data source. If such synchronization is necessary then a default implementation w/o sync can be a risk.

Reply:
LocalCacheManager uses a write lock to prevent each thread from loading its own data. so the same data can be loaded only once
The other cache managers do not have locks mechanism and are more for testing instead of production use, so the implementation is the same (without synchronization one).

This code will not introduce bugs, but just inefficiently with duplicate works.

The problem here is that because it contains the default impl, if users impl a new production cache manager with synchronization mechanism, it may forget to implement this method in a more efficient approach.....

let me see how to deal with it

Reply:
Right. I tend to have default implementations for interface methods only if it's absolutely obvious that what they do is the "default" in any sense (like wrapping a `byte[]` in a `ByteBuffer`, or just throwing an `UnimplementedException`). More complex logics in default implementations can lead to surprising bugs.

Reply:
Totally make sense!